toolchain/driver/clang_runner_test.cpp

```diff
@@ -0,0 +1,150 @@
+using ::testing::internal::GetCapturedStdout;
+
+// Calls the provided lambda with `stderr` and `stdout` captured and saved into
+// the provided output parameters. The lambda's result is returned. It is
+// important to not put anything inside the lambda whose output would be useful
+// in interpreting test errors such as Google Test assertions as their output
+// will end up captured as well.
+template <typename CallableT>
+static auto RunWithCapturedOutput(std::string& out, std::string& err,
+                                  CallableT callable) {
+  CaptureStderr();
+  CaptureStdout();
+  auto result = callable();
+  // No need to flush stderr.
+  err = GetCapturedStderr();
+  llvm::outs().flush();
+  out = GetCapturedStdout();
+  return result;
+}
+
+TEST(ClangRunnerTest, Version) {
+  TestRawOstream test_os;
+  std::string target = llvm::sys::getDefaultTargetTriple();
+  ClangRunner runner("./toolchain/driver/run_clang_test", target, &test_os);
+
+  std::string out;
+  std::string err;
+  EXPECT_TRUE(RunWithCapturedOutput(out, err,
+                                    [&] { return runner.Run({"--version"}); }));
+  // The arguments to Clang should be part of the verbose log.
+  EXPECT_THAT(test_os.TakeStr(), HasSubstr("--version"));
+
+  // No need to flush stderr, just check its contents.
+  EXPECT_THAT(err, StrEq(""));
+
+  // Flush and get the captured stdout to test that this command worked.
+  // We don't care about any particular version, just that it is printed.
+  EXPECT_THAT(out, HasSubstr("clang version"));
+  // The target should match what we provided.
+  EXPECT_THAT(out, HasSubstr((llvm::Twine("Target: ") + target).str()));
+  // The installation should come from the above path of the test binary.
+  EXPECT_THAT(out, HasSubstr("InstalledDir: ./toolchain/driver"));
+}
+
+// Utility to write a test file. We don't need the full power provided here yet,
+// but we anticipate adding more tests such as compiling basic C++ code in the
+// future and this provides a basis for building those tests.
+static auto WriteTestFile(llvm::StringRef name_suffix, llvm::Twine contents)
+    -> std::filesystem::path {
+  std::filesystem::path test_tmpdir;
+  if (char* tmpdir_env = getenv("TEST_TMPDIR"); tmpdir_env != nullptr) {
+    test_tmpdir = std::string(tmpdir_env);
+  } else {
+    test_tmpdir = std::filesystem::temp_directory_path();
+  }
+
+  const auto* unit_test = ::testing::UnitTest::GetInstance();
+  const auto* test_info = unit_test->current_test_info();
+  std::filesystem::path test_file =
+      test_tmpdir / llvm::formatv("{0}_{1}_{2}", test_info->test_suite_name(),
+                                  test_info->name(), name_suffix)
+                        .str();
+  // Make debugging a bit easier by cleaning up any files from previous runs.
+  // This is only necessary when not run in Bazel's test environment.
+  std::filesystem::remove(test_file);
+  CARBON_CHECK(!std::filesystem::exists(test_file));
```

Comment:
Had you considered using gtest's teardown to remove created files, rather than deleting name collisions this way? You might then also be able to create a simpler filename, rather than the complex formula.

Note this may also be masking a duplicate "foo.o" mistake below.

Reply:
> However, I don't quite see how using teardown would help with the complex temporary formula... I added this in another test and I thought this was needed in part because of test sharding and such.... But maybe I'm mis-remembering.

I've never heard of this issue with sharding: each shard gets its own tmpdir.

That said, you might consider something like `llvm::sys::fs::createUniqueDirectory(test_tmpdir / test_info->test_suite_name(), result_path)`, then you shouldn't need to remove files and would still have files available after execution.

https://llvm.org/doxygen/namespacellvm_1_1sys_1_1fs.html#abf677459f5137cfb49f6d81b27ea54a1